### ISSUE TITLE
ConversationalRetrievalQAChain with external memory

### ISSUE THREAD

**Issue by dstoyanoff** [langchain-ai/langchainjs#1607]
Hello,
I am having trouble making that work and would appreciate any help. Here is a snipped that uses the BufferMemory and it does work correctly:

```
const model = new ChatOpenAI({
    modelName: "gpt-3.5-turbo",
    openAIApiKey: process.env.OPENAI_API_KEY,
  });

  const chain = ConversationalRetrievalQAChain.fromLLM(model, new Retriever(), {
    memory: new BufferMemory({
      memoryKey: "chat_history",
      inputKey: "question",
      outputKey: "text",
      returnMessages: true,
    }),
    returnSourceDocuments: true,
  });

  const question = "How do I start a personal brand?";

  const res = await chain.call({
    question,
  });

  console.dir({ res }, { depth: null });

  const followUpRes = await chain.call({
    question: "Any other tips?",
  });

  console.dir({ followUpRes }, { depth: null });
```

I can see that the follow-up question was actually converted to `What are some additional tips for building a personal brand?`.

The following snipper however uses external memory and I can't make it work:
```
const model = new ChatOpenAI({
    modelName: "gpt-3.5-turbo",
    openAIApiKey: process.env.OPENAI_API_KEY,
  });

  const pastMessages = [
    new HumanChatMessage("How do I start a personal brand?"),
    new AIChatMessage(
      "There are many ways to start a personal brand, but some initial steps you could take include identifying your niche or area of expertise, developing a strong online presence through social media and/or a website, creating and sharing content that showcases your skills or values, and networking with others in your industry or community. It's also important to be authentic and consistent in your messaging and branding efforts",
    ),
  ];

  const memory = new BufferMemory({
    chatHistory: new ChatMessageHistory(pastMessages),
  });

  const chain = ConversationalRetrievalQAChain.fromLLM(model, new Retriever(), {
    returnSourceDocuments: true,
    memory,
  });

  const followUpRes = await chain.call({
    question: "Any other tips?",
  });

  console.dir({ followUpRes }, { depth: null });
```

The error that I get is `throw new Error(`Chat history key ${this.chatHistoryKey} not found.`);`. I was able to make it work using `chat_history` and passing `question + res.text`, but since my use-case might have more than 1 question/answer pair, that looks weird.

Thanks

**Comment by duyhnguyen216** [comment]
when declaring your memory object, try adding the memory key.

```
const memory = new BufferMemory({
            chatHistory: new ChatMessageHistory(chatMessages),
            returnMessages: true, //optional
            memoryKey: "chat_history",
        });
```

**Comment by dstoyanoff** [comment]
When I do that, I start getting this error:
```
Error: input values have 2 keys, you must specify an input key or pass only 1 key as input
```

**Comment by duyhnguyen216** [comment]
> const chain = ConversationalRetrievalQAChain.fromLLM(model, new Retriever(), {
>     returnSourceDocuments: true,
>     memory,
>   });

in 

```
const chain = ConversationalRetrievalQAChain.fromLLM(model, new Retriever(), {
    returnSourceDocuments: true,
    memory,
  });

```
set `returnSourceDocuments` to `false` or specify the input key

**Comment by dstoyanoff** [comment]
Setting to false does work, but I still need the documents. What should the input key be set to? I see it's set to "question" by default and my input is also called question, but I see the same error from above. 
Thanks

**Comment by muhammadfaizan** [comment]
@dstoyanoff  this will resolve your issue

```ts
const memory = new BufferMemory({
    chatHistory: new ChatMessageHistory(history),
    memoryKey: 'chat_history'
  })
  ```

**Comment by dstoyanoff** [comment]
@muhammadfaizan, this is exactly what i have now and as stated above, it doesn't work

**Comment by carlosveloso** [comment]
I've been dealing with the same problem for the past couple of hours. I believe it's an issue of how to handle the inputKeys and the outputKeys that are needed in some cases. Here is the corrected version of your code:
```js

const chatHistory = new ChatMessageHistory(pastMessages);

const memory = new BufferMemory({
  chatHistory,
  inputKey: "memoryKey",
  outputKey: "text",
});

const chain = ConversationalRetrievalQAChain.fromLLM(
  model,
  new Retriever(),
  {
    returnSourceDocuments: true,
    inputKey: "question",
    memory,
  }
);

const followUpRes = await chain.call({
  question: "Any other tips?",
  chat_history: chatHistory,
});,
  {
    returnSourceDocuments: true,
    inputKey: "question",
    memory,
  }
);

const followUpRes = await chain.call({
  question: "Any other tips?",
  chat_history: chatHistory,
});
```

Also, sometimes the error message isn't clear because if it's an outputKey issue, it displays an input key error. This happens because it uses the getInputValue method to check for the outputValues. 😅https://github.com/hwchase17/langchainjs/blob/b696686744231bf2b383d3f4e17375c12bd46b76/langchain/src/memory/chat_memory.ts#L43

**Comment by joaopcm** [comment]
@carlosveloso, I'm fixing the issue with the wrong error message here -> https://github.com/hwchase17/langchainjs/pull/1778

**Comment by satpalsr** [comment]
@carlosveloso Is this a typo at end?

```
const followUpRes = await chain.call({
  question: "Any other tips?",
  chat_history: chatHistory,
});, // This got closed.

// What's this part?
  {
    returnSourceDocuments: true,
    inputKey: "question",
    memory,
  }
);

// This seems ok
const followUpRes = await chain.call({
  question: "Any other tips?",
  chat_history: chatHistory,
});
```

**Comment by cdiddy77** [comment]
I was able to get what I needed (retrieval + chat history). I ended up not using ConversationalRetrievalQAChain but just regular RetrievalQAChain. I guess I just don't really understand how CRQAC is supposed to work. Hopefully this makes some sense:

```javascript
 const vectorStore = // ... some vectorstore
 const chatPrompt = ChatPromptTemplate.fromPromptMessages([
      SystemMessagePromptTemplate.fromTemplate(
        "You are <WHATEVER KIND OF AI> "
      ),
      new MessagesPlaceholder("history"),
      HumanMessagePromptTemplate.fromTemplate("{question}"),
    ]);

    const chat = new ChatOpenAI({
      temperature: 0.9,
      modelName: "gpt-3.5-turbo",
      maxTokens: 256,
    });

    const chain = RetrievalQAChain.fromLLM(chat, vectorStore.asRetriever(), {
           returnSourceDocuments: true,
    });

    const chatHistory = new ChatMessageHistory();
    chain.memory = new BufferMemory({
      returnMessages: true,
      memoryKey: "history",
      chatHistory,
    });

    let continueChat = true;
    while (continueChat) {
      const promptResult = // ... get user question
      if (promptResult === "exit") {
        continueChat = false;
      } else {
        const response = await chain.call({
          query: await chatPrompt.format({
            question: promptResult,
            history: await chatHistory.getMessages(),
          }),
        });
        // pass along response to user
        // console.log(chalk.cyan(`it: ${response.text}`));
      }
    }

```

**Comment by joaopcm** [comment]
I had to deal with the same issue. I ended up opening an issue explaining why the issue occurs and how I got it fixed [here](https://github.com/hwchase17/langchainjs/issues/1800).

I also [opened a PR](https://github.com/hwchase17/langchainjs/pull/1801) with the fix. You guys can review it and see if it makes sense to you.

**Comment by araujooj** [comment]
Hey, guys - same error here. Do you have any updates on this?

My code: 

utils.ts
```ts
type ChatMessage = {
  role: 'user' | 'assistant'
  content: string
}

export const getChatMessages = (history: ChatMessage[]) => {
  return history.map((message) =>
    message.role === 'user'
      ? new HumanMessage(message.content)
      : new AIMessage(message.content),
  )
}

export const extractLastQuestion = (messages: ChatMessage[]) => {
  const question =
    messages.length > 0 ? messages[messages.length - 1].content : ''
  const previousMessages = messages.slice(0, messages.length - 1)

  return { question, previousMessages }
}

export const getMemory = (messages: ChatMessage[]) => {
  const { question, previousMessages } = extractLastQuestion(messages)

  const messageHistory = getChatMessages(previousMessages)

  const memory = new BufferMemory({
    memoryKey: 'chat_history',
    inputKey: 'input',
    outputKey: 'output',
    chatHistory: new ChatMessageHistory(messageHistory),
  })

  return { memory, question }
}
```

route.ts
```ts
export async function POST(req: Request) {
  const { messages } = await req.json()
  const { memory, question } = getMemory(messages) 

  const { stream, handlers } = LangChainStream()

  const template = `
  You are a chatbot helping humans with their questions. Related to a fantasy universe called "Nove Círculos".
  You should always answers on: pt-BR
  
  {context}
  
  Human: {question}
  Assistant:
  `

  const prompt = new PromptTemplate({
    template,
    inputVariables: ['question', 'context'],
  })

  const streamingModel = new ChatOpenAI({
    streaming: true,
    modelName: 'gpt-3.5-turbo-16k',
    callbacks: [handlers],
  })

  const nonStreamingModel = new ChatOpenAI({
    modelName: 'gpt-4',
  })

  const retriever = new ChatGPTPluginRetriever({
    url: process.env.RETRIEVER_URL as string,
    auth: {
      bearer: process.env.PLUGIN_BEARER_TOKEN as string,
    },
    topK: 10,
  })

  const CONDENSE_QUESTION_TEMPLATE = `
  Given the following conversation and a follow up input, if it is a question rephrase it to be a standalone question.
  If it is not a question, just summarize the message.
  
  Chat history:
  {chat_history}
  Follow up input: {question}
  Standalone input:
  `

  const chain = ConversationalRetrievalQAChain.fromLLM(
    streamingModel,
    retriever,
    {
      memory,
      returnSourceDocuments: true,
      qaChainOptions: {
        type: 'stuff',
        prompt: prompt,
      },
      questionGeneratorChainOptions: {
        llm: nonStreamingModel,
        template: CONDENSE_QUESTION_TEMPLATE,
      },
    },
  )

  chain.call({ question })

  return new StreamingTextResponse(stream)
}
```

**Comment by MKorostoff** [comment]
I was able to get this working by doing something very similar to the suggestion by @carlosveloso, thanks for the tip!

```ts
const chain = ConversationalRetrievalQAChain.fromLLM(
  model,
  vectorStore.asRetriever(undefined),
  {
    memory: new BufferMemory({
      memoryKey: "chat_history",
      inputKey: "question",
      outputKey: "text"
    }),
    returnSourceDocuments: true,
  }
);

const generatednswer = await chain.call({
  question: chatMessage
});
```

**Comment by carlosveloso** [comment]
You should be able to pass the your `pastMessages` array like this:
```ts
import { ChatMessageHistory } from 'langchain/memory';

const messageHistory = new ChatMessageHistory(pastMessages);

const memory = new BufferMemory({
  memoryKey: "chat_history",
  inputKey: "question",
  outputKey: "text",
  chatHistory: messageHistory,
});
```
 (disclaimer, I haven't used this in a while but it should work)

> > I was able to get this working by doing something very similar to the suggestion by @carlosveloso, thanks for the tip!
> > ```ts
> > const chain = ConversationalRetrievalQAChain.fromLLM(
> >   model,
> >   vectorStore.asRetriever(undefined),
> >   {
> >     memory: new BufferMemory({
> >       memoryKey: "chat_history",
> >       inputKey: "question",
> >       outputKey: "text"
> >     }),
> >     returnSourceDocuments: true,
> >   }
> > );
> > 
> > const generatednswer = await chain.call({
> >   question: chatMessage
> > });
> > ```
> 
> Where did you pass the historical messages though?
> 
> For example,
> 
> Here are my historical messages:
> 
> ```
> 
>     const pastMessages = messages
>       .slice(0, messages.length - 1)
>       .map((message) => {
>         if (message.role === 'user') {
>           return new HumanMessage(message.content);
>         } else if (message.role === 'assistant' || message.role === 'system') {
>           return new AIMessage(message.content);
>         }
>       });
>     
>     const chatHistory = pastMessages.map((message) => message.toString()).join('\n');
>     
> ```
